frame/contracts/src/wasm/runtime.rs

```diff
@@ -632,6 +648,46 @@ where
 		}
 	}
 
+	/// Extracts the size of the overwritten value or `u32::MAX`.
+	fn overwritten_len(outcome: WriteOutcome) -> u32 {
+		match outcome {
+			// We cannot use `0` as sentinel value because there could be a zero sized
+			// storage entry which is different from a non existing one.
+			WriteOutcome::New => u32::MAX,
+			WriteOutcome::Overwritten(len) => len,
+			WriteOutcome::Taken(value) => value.len() as u32,
+		}
+	}
+
+	fn set_storage(
+		&mut self,
+		key_ptr: u32,
+		value_ptr: u32,
+		value_len: u32,
+	) -> Result<u32, TrapReason> {
+		self.charge_gas(RuntimeCosts::SetStorage(value_len))?;
+		if value_len > self.ext.max_value_size() {
+			Err(Error::<E::T>::ValueTooLarge)?;
+		}
+		let mut key: StorageKey = [0; 32];
+		self.read_sandbox_memory_into_buf(key_ptr, &mut key)?;
+		let value = Some(self.read_sandbox_memory(value_ptr, value_len)?);
+		self.ext
+			.set_storage(key, value, false)
```

Comment:
Because we are in `seal_set_storage` which should just overwrite the value without reading (taking) the existing value. This is what `seal_take_storage` is for.